linera-rpc/src/grpc/server.rs

```diff
@@ -399,75 +401,133 @@ where
+                {
+                    warn!("Dropped 1 cross-chain message intentionally.");
+                    return;
+                }
+
+                let mut i = 0;
+                while i < cross_chain_max_retries {
+                    linera_base::time::timer::sleep(cross_chain_sender_delay).await;
+                    // Delay increases linearly with the attempt number.
+                    if i > 0 {
+                        _permit = None;
+                        linera_base::time::timer::sleep(cross_chain_retry_delay * i).await;
+                        _permit = Some(semaphore_clone.clone().acquire_owned().await.unwrap());
                     }
 
-                    for i in 0..cross_chain_max_retries {
-                        // Delay increases linearly with the attempt number.
-                        linera_base::time::timer::sleep(
-                            cross_chain_sender_delay + cross_chain_retry_delay * i,
-                        )
-                        .await;
-
-                        let result = || async {
-                            let cross_chain_request = cross_chain_request.clone().try_into()?;
-                            let request = Request::new(cross_chain_request);
-                            let mut client =
-                                ValidatorWorkerClient::new(pool.channel(remote_address.clone())?)
-                                    .max_encoding_message_size(GRPC_MAX_MESSAGE_SIZE)
-                                    .max_decoding_message_size(GRPC_MAX_MESSAGE_SIZE);
-                            let response = client.handle_cross_chain_request(request).await?;
-                            Ok::<_, anyhow::Error>(response)
-                        };
-                        match result().await {
-                            Err(error) => {
-                                warn!(
-                                    nickname,
-                                    %error,
-                                    i,
-                                    from_shard = this_shard,
-                                    to_shard = shard_id,
-                                    "Failed to send cross-chain query",
-                                );
+                    let result = || async {
+                        let cross_chain_request = cross_chain_request.clone().try_into()?;
+                        let request = Request::new(cross_chain_request);
+                        let mut client =
+                            ValidatorWorkerClient::new(pool.channel(remote_address.clone())?)
+                                .max_encoding_message_size(GRPC_MAX_MESSAGE_SIZE)
+                                .max_decoding_message_size(GRPC_MAX_MESSAGE_SIZE);
+                        let response = client.handle_cross_chain_request(request).await?;
+                        Ok::<_, anyhow::Error>(response)
+                    };
+                    match result().await {
```

Comment:
I think if you await `result` above then swap the order of the matches here you can avoid some duplication?

Reply:
Like this: https://github.com/Twey/linera-protocol/commit/379221680027e6d82cb227d9100e0df2d58afe1d

Reply:
Thanks! Feel free to push!